crates/biome_cli/src/commands/check.rs

```diff
@@ -43,174 +31,128 @@ pub(crate) struct CheckCommandPayload {
+            .formatter
+            .get_or_insert_with(PartialFormatterConfiguration::default);
+
+        if self.formatter_enabled.is_some() {
+            formatter.enabled = self.formatter_enabled;
+        }
 
-    if linter_enabled.is_some() {
-        linter.enabled = linter_enabled;
-    }
+        let linter = fs_configuration
+            .linter
+            .get_or_insert_with(PartialLinterConfiguration::default);
 
-    let organize_imports = fs_configuration
-        .organize_imports
-        .get_or_insert_with(PartialOrganizeImports::default);
+        if self.linter_enabled.is_some() {
+            linter.enabled = self.linter_enabled;
+        }
 
-    if organize_imports_enabled.is_some() {
-        organize_imports.enabled = organize_imports_enabled;
-    }
+        let organize_imports = fs_configuration
+            .organize_imports
+            .get_or_insert_with(PartialOrganizeImports::default);
 
-    let assists = fs_configuration
-        .assists
-        .get_or_insert_with(PartialAssistsConfiguration::default);
+        if self.organize_imports_enabled.is_some() {
+            organize_imports.enabled = self.organize_imports_enabled;
+        }
+
+        let assists = fs_configuration
+            .assists
+            .get_or_insert_with(PartialAssistsConfiguration::default);
 
-    if assists_enabled.is_some() {
-        assists.enabled = assists_enabled;
+        if self.assists_enabled.is_some() {
+            assists.enabled = self.assists_enabled;
+        }
+
+        if let Some(mut configuration) = self.configuration.clone() {
+            if let Some(linter) = configuration.linter.as_mut() {
+                // Don't overwrite rules from the CLI configuration.
+                // Otherwise, rules that are disabled in the config file might
+                // become re-enabled due to the defaults included in the CLI
+                // configuration.
+                linter.rules = None;
+            }
+            fs_configuration.merge_with(configuration);
+        }
+
+        Ok(fs_configuration)
     }
 
-    if let Some(mut configuration) = configuration {
-        if let Some(linter) = configuration.linter.as_mut() {
-            // Don't overwrite rules from the CLI configuration.
-            // Otherwise, rules that are disabled in the config file might
-            // become re-enabled due to the defaults included in the CLI
-            // configuration.
-            linter.rules = None;
+    fn get_files_to_process(
+        &self,
+        fs: &DynRef<'_, dyn FileSystem>,
+        configuration: &PartialConfiguration,
+    ) -> Result<Vec<OsString>, CliDiagnostic> {
+        if let Some(paths) = get_files_to_process(
+            self.since.as_deref(),
+            self.changed,
+            self.staged,
+            fs,
+            configuration,
+        )? {
+            Ok(paths)
+        } else {
+            Ok(self.paths.clone())
         }
```

Comment:
It feels a little bit confusing to have the `get_files_to_process()` method and a utility function of the same name. I think I would suggest renaming the utility (maybe `get_files_to_process_with_ci_options`?). Syntax-wise, I think the readability of these can be improved a bit by moving the function call with all its arguments (and `?` operator) out of the `if` condition:

```suggestion
        let paths = get_files_to_process_with_ci_options(
            self.since.as_deref(),
            self.changed,
            self.staged,
            fs,
            configuration,
        )?.ok_or_else(|| self.paths.clone());
        Ok(paths)
```